src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.TranslationAction_Actions.cs

```diff
@@ -57,6 +57,59 @@ public override GeneratorDriver TransformGeneratorDriver(GeneratorDriver generat
             }
         }
 
+        internal sealed class TouchDocumentsAction : TranslationAction
+        {
+            private readonly ImmutableArray<DocumentState> _newStates;
+
+            private TouchDocumentsAction(
+                ProjectState oldProjectState,
+                ProjectState newProjectState,
+                ImmutableArray<DocumentState> newStates)
+                : base(oldProjectState, newProjectState)
+            {
+                _newStates = newStates;
+            }
+
+            public static TranslationAction Create(
+                ProjectState oldProjectState,
+                ProjectState newProjectState,
+                ImmutableArray<DocumentState> newStates)
+            {
+                // Special case when we're only updating a single document.  This case can be optimized more, and
+                // corresponds to the common case of a single file being edited.
+                return newStates.Length == 1
+                    ? new TouchDocumentAction(oldProjectState, newProjectState, oldProjectState.DocumentStates.GetRequiredState(newStates[0].Id), newStates[0])
+                    : new TouchDocumentsAction(oldProjectState, newProjectState, newStates);
+            }
+
+            public override async Task<Compilation> TransformCompilationAsync(Compilation oldCompilation, CancellationToken cancellationToken)
+            {
+                var finalCompilation = oldCompilation;
+                for (int i = 0, n = _newStates.Length; i < n; i++)
+                {
+                    cancellationToken.ThrowIfCancellationRequested();
+                    var newState = _newStates[i];
+                    var oldState = this.OldProjectState.DocumentStates.GetRequiredState(newState.Id);
+                    finalCompilation = finalCompilation.ReplaceSyntaxTree(
+                        await oldState.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false),
+                        await newState.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false));
+                }
+
+                return finalCompilation;
+            }
+
+            /// <inheritdoc cref="TouchDocumentAction.CanUpdateCompilationWithStaleGeneratedTreesIfGeneratorsGiveSameOutput"/>
+            public override bool CanUpdateCompilationWithStaleGeneratedTreesIfGeneratorsGiveSameOutput
+                => true;
```

Comment:
note 100% sure about this @jasonmalinowski .  however, as it is true for TouchDocumentAction, i don't know why it would not be true here just because we're touching multiple docs.